presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriver.java

```diff
@@ -57,31 +79,332 @@
+     * QueryExecutors regardless of their refcount by calling the cache's
+     * close() method, and in doing so closes their HTTP clients. If anybody is
+     * still trying to use a PrestoConnection created by the driver, they're
+     * SOL.
+     */
+    private static class ReferenceCountingLoadingCache<K, V>
+    {
+        private class Holder
+        {
+            private final V value;
+            private int refcount;
+            private int cleanerCount;
+            private ScheduledFuture oldCleanup;
+
+            private Holder(V value)
+            {
+                this.value = requireNonNull(value);
+            }
+
+            private V get()
+            {
+                return value;
+            }
+
+            private int getRefcount()
+            {
+                return refcount;
+            }
+
+            private int getCleanerCount()
+            {
+                return cleanerCount;
+            }
+
+            private int reference()
+            {
+                setCleanup(null);
+                return ++refcount;
+            }
+
+            private int dereference()
+            {
+                return --refcount;
+            }
+
+            private void setCleanup(ScheduledFuture newCleanup)
+            {
+                if (oldCleanup != null && oldCleanup.cancel(false)) {
+                    --cleanerCount;
+                }
+                oldCleanup = newCleanup;
+            }
+
+            private void scheduleCleanup(
+                    ScheduledExecutorService cleanupService,
+                    Runnable cleanupTask,
+                    long delay,
+                    TimeUnit unit)
+            {
+                checkState(refcount == 0, "non-zero refcount in scheduleCleanup");
+                ++cleanerCount;
+                ScheduledFuture newCleanup = cleanupService.schedule(cleanupTask, delay, unit);
+                setCleanup(newCleanup);
+            }
+
+            private void releaseForCleaning()
+            {
+                --cleanerCount;
+            }
+        }
+
+        private final ScheduledExecutorService valueCleanupService;
+        private final LoadingCache<K, Holder> backingCache;
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+
+        private ReferenceCountingLoadingCache(CacheLoader<K, V> loader, Consumer<V> disposer)
+        {
+            this.valueCleanupService = new ScheduledThreadPoolExecutor(1, daemonThreadsNamed("cache-cleanup-%s"));
+            this.backingCache = CacheBuilder.newBuilder()
+                    .removalListener(new RemovalListener<K, Holder>() {
+                        @Override
+                        public void onRemoval(RemovalNotification<K, Holder> notification)
+                        {
+                            Holder holder = notification.getValue();
+                            /*
+                             * The docs say that both the key and value may be
```

Comment:
Docs here: http://google.github.io/guava/releases/19.0/api/docs/com/google/common/cache/RemovalNotification.html

Anybody got a better interpretation? `QueryExecutor` overrides `finalize()`, so technically it's OK if it gets garbage collected before we get a `RemovalNotification`. I'd still like to be able to have the sanity check here.